staging/src/k8s.io/pod-security-admission/policy/checks.go

```diff
@@ -0,0 +1,141 @@
+	Name() string
+	// Description returns markdown, used for the body of docs
+	Description() string
+	// Delta describes changes since the previous version of this check.
+	// Optional, should be empty if there was no previous version of this check
+	Delta() string
+}
+
+type doc struct {
+	name        string
+	description string
+	delta       string
+}
+
+func (d doc) Name() string        { return d.name }
+func (d doc) Description() string { return d.description }
+func (d doc) Delta() string       { return d.delta }
+
+type check struct {
+	id string
+	doc
+	checkPod func(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult
+}
+
+func (c *check) CheckPod(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
+	return c.checkPod(podMetadata, podSpec)
+}
+func (c *check) ID() string {
+	return c.id
+}
+
+// AggergateCheckResult holds the aggregate result of running CheckPod across multiple checks.
+type AggregateCheckResult struct {
+	// Allowed indicates if all checks allowed the pod.
+	Allowed bool
+	// ForbiddenReasons is a slice of the forbidden reasons from all the forbidden checks. It should not include empty strings.
+	// ForbiddenReasons and ForbiddenDetails must have the same number of elements, and the indexes are for the same check.
+	ForbiddenReasons []string
+	// ForbiddenDetails is a slice of the forbidden details from all the forbidden checks. It may include empty strings.
+	// ForbiddenReasons and ForbiddenDetails must have the same number of elements, and the indexes are for the same check.
+	ForbiddenDetails []string
+}
+
+// ForbiddenReason returns a comma-separated string of of the forbidden reasons.
+// Example: host ports, privileged containers, non-default capabilities
+func (a *AggregateCheckResult) ForbiddenReason() string {
+	return strings.Join(a.ForbiddenReasons, ", ")
+}
+
+// ForbiddenDetail returns a detailed forbidden message, with non-empty details formatted in
+// parentheses with the associated reason.
+// Example: host ports (8080, 9090), privileged containers, non-default capabilities (NET_RAW)
+func (a *AggregateCheckResult) ForbiddenDetail() string {
+	var b strings.Builder
+	for i := 0; i < len(a.ForbiddenReasons); i++ {
+		b.WriteString(a.ForbiddenReasons[i])
+		if a.ForbiddenDetails[i] != "" {
+			b.WriteString(" (")
+			b.WriteString(a.ForbiddenDetails[i])
+			b.WriteString(")")
+		}
+		if i != len(a.ForbiddenReasons)-1 {
+			b.WriteString(", ")
+		}
+	}
+	return b.String()
+}
+
+// AggregateCheckPod runs all the checks and aggregates the forbidden results into a single CheckResult.
+// The aggregated reason is a comma-separated
+func AggregateCheckPod(checks []Check, podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) AggregateCheckResult {
+	var (
+		reasons []string
+		details []string
+	)
+	for _, check := range checks {
+		r := check.CheckPod(podMetadata, podSpec)
+		if !r.Allowed {
+			reasons = append(reasons, r.ForbiddenReason)
```

Comment:
empties look like they'll produce weird output.

Reply:
can fail if we encounter empty strings (or do ***MISSING*** or something and check for that in tests)